qa/tasks/ceph2.py

```diff
@@ -0,0 +1,700 @@
+                    logger=log.getChild(mon),
+                    wait=False,
+                    started=True,
+                )
+
+                while True:
+                    log.info('Waiting for %d mons in monmap...' % (num_mons))
+                    r = shell(
+                        ctx=ctx,
+                        cluster_name=cluster_name,
+                        remote=remote,
+                        args=[
+                            'ceph', 'mon', 'dump', '-f', 'json',
+                        ],
+                        stdout=StringIO(),
+                    )
+                    j = json.loads(r.stdout.getvalue())
+                    if len(j['mons']) == num_mons:
+                        break
+                    time.sleep(1)
+
+        ## FIXME: refresh ceph.conf files for all mons + first mgr ##
+
+        yield
+
+    finally:
+        pass
+
+@contextlib.contextmanager
+def ceph_mgrs(ctx, config):
+    """
+    Deploy any additional mgrs
+    """
+    cluster_name = config['cluster']
+    fsid = ctx.ceph[cluster_name].fsid
+    testdir = teuthology.get_testdir(ctx)
+
+    try:
+        nodes = []
+        for remote, roles in ctx.cluster.remotes.items():
+            for mgr in [r for r in roles
+                        if teuthology.is_type('mgr', cluster_name)(r)]:
+                c_, _, id_ = teuthology.split_role(mgr)
+                if c_ == cluster_name and id_ == ctx.ceph[cluster_name].first_mgr:
+                    continue
+                log.info('Adding %s on %s' % (mgr, remote.shortname))
+                nodes.append(remote.shortname + '=' + id_)
+
+                ctx.daemons.register_daemon(
+                    remote, 'mgr', id_,
+                    cluster=cluster_name,
+                    fsid=fsid,
+                    logger=log.getChild(mgr),
+                    wait=False,
+                    started=True,
+                )
+
+        shell(ctx, cluster_name, remote, [
+            'ceph', 'orchestrator', 'mgr', 'update',
+            str(len(nodes) + 1)] + nodes
+        )
+
+        yield
+
+    finally:
+        pass
+
+@contextlib.contextmanager
+def ceph_osds(ctx, config):
+    """
+    Deploy OSDs
+    """
+    cluster_name = config['cluster']
+    fsid = ctx.ceph[cluster_name].fsid
+    try:
+        log.info('Deploying OSDs...')
+
+        # provision OSDs in order
+        id_to_remote = {}
+        devs_by_remote = {}
+        for remote, roles in ctx.cluster.remotes.items():
+            devs_by_remote[remote] = teuthology.get_scratch_devices(remote)
+            for osd in [r for r in roles
+                        if teuthology.is_type('osd', cluster_name)(r)]:
+                _, _, id_ = teuthology.split_role(osd)
+                id_to_remote[id_] = (osd, remote)
+
+        cur = 0
+        for osd_id, t in id_to_remote.items():
+            osd, remote = t
+            _, _, id_ = teuthology.split_role(osd)
+            assert int(id_) == cur
```

Comment:
looks like this is requiring the config to list osds in increasing order (and assumes hosts are ordered the same as in the yaml file). It'd be nice to do the provisioning here in a way that didn't require that in the future, e.g. looping through the osds in id order rather than config file order

Reply:
yeah i didn't do this quite right.  the intent is to put it in the dict so i could sort the keys numerically, so that we always provision the osd ids in order.